src/proxy.js

```diff
@@ -0,0 +1,278 @@
+     * @return {object} represents information about proxy.
+     */
+    getInfo(sessionId) {
+        const { groups, country, password, port, hostname } = this;
+        const url = this.getUrl(sessionId);
+
+        return {
+            sessionId,
+            url,
+            groups,
+            country,
+            password,
+            hostname,
+            port,
+        };
+    }
+
+    /**
+     * Returns proxy url.
+     * @param {string} sessionId
+     * Apify Proxy session identifier to be used with requests.
+     * All HTTP requests going through the proxy with the same session identifier
+     * will use the same target proxy server (i.e. the same IP address).
+     * The identifier can only contain the following characters: `0-9`, `a-z`, `A-Z`, `"."`, `"_"` and `"~"`.
+     * @return {string} - the proxy url
+     */
+    getUrl(sessionId) {
+        const username = this._getUsername(sessionId);
+        const { password, hostname, port } = this;
+
+        return `${PROTOCOL}://${username}:${password}@${hostname}:${port}`;
+    }
+
+    /**
+     * Returns proxy username.
+     * @return {string} - the proxy username
+     * @param {string} sessionId
+     * @private
+     */
+    _getUsername(sessionId) {
+        let username;
+        const { groups, country } = this;
+        if (groups || sessionId || country) {
+            const parts = [];
+
+            if (groups && groups.length) {
+                parts.push(`groups-${groups.join('+')}`);
+            }
+            if (sessionId) {
+                parts.push(`session-${sessionId}`);
+            }
+            if (country) {
+                parts.push(`country-${country}`);
+            }
+
+            username = parts.join(',');
+        } else {
+            username = 'auto';
+        }
+        return username;
+    }
+
+    /**
+     * Checks if apify token is provided in env
+     * and gets the password via API and sets it to env
+     * @returns {Promise<void>}
+     * @private
+     */
+    async _setPasswordIfToken() {
+        const token = process.env[ENV_VARS.TOKEN] || LOCAL_ENV_VARS[ENV_VARS.TOKEN];
+        if (!this.password) {
+            if (token) {
+                const { proxy: { password, groups } } = await apifyClient.users.getUser({ token, userId: 'me' });
+                this.password = password;
+                this.availableGroups = groups;
+            } else {
+                throw new Error(this._getMissingParamErrorMgs('password', ENV_VARS.PROXY_PASSWORD));
+            }
+        }
+    }
+
+    /**
+     * Checks the status of apify proxy and
+     * throws an access denied error if the status is not "connected".
+     * Also checks if all proxy groups are available if token is provided.
+     * @returns {Promise<void>}
+     * @private
+     */
+    async _checkAccess() {
+        // Check if user has access to selected proxy groups
+        await this._checkAccessToProxyGroups();
```

Comment:
BTW this check is performed automatically by the proxy status page, no need to do it here again...

Reply:
@jancurn The idea of this check was to give a user a better error message with the exact wrong group name. Otherwise, you get just a message that you do not have access to Apify Proxy...

Reply:
proxy.apify.com?format=json returns this info too:

```
{"connected":false,"connectionError":"Invalid username: proxy group &quot;BOB&quot; not found or not accessible.","username":"groups-BOB","userId":"some_user_id","passwordValid":true,"rejectUrl":"http://proxy.apify.com/?reject=groups-BOB"}
```

Reply:
Ok, fair...but I tried it with a real proxy name that I do not have access to and it returns just its ID which could be a bit confusing for a user. Is it a bug?

`{"connected":false,"connectionError":"Invalid username: proxy group &quot;A7hH8X6AAMWothbqA&quot; not found or not accessible.","username":"groups-CZECH_LUMINATI+FRANCE","userId":"user-id","passwordValid":true,"rejectUrl":"http://proxy.apify.com/?reject=groups-CZECH_LUMINATI%2BFRANCE"}`

Reply:
Oh, interesting, are you passing the groupId or name? The status should return a human-friendly error

Reply:
I was passing group names e.g. `groups-CZECH_LUMINATI%2BFRANCE`

Reply:
Oh you're right, the error message is bad in case the group is not accessible, we'll fix that.

```
Connection error / Invalid username: proxy group "MftBSp3XeEEYtHzH4" not found or not accessible
```

Reply:
Ok the PR with this change to Apify Proxy is ready

Reply:
Ok. Perfect thanks! :)